text/0000-subslice-pattern-syntax.md

```diff
@@ -0,0 +1,207 @@
+on `..`.  
+The two simplest variations are `..PAT` and `PAT..`.
+
+#### Ambiguity
+
+The issue is that these syntaxes are ambiguous with half-bounded ranges `..END` and `BEGIN..`.  
+To be precise, such ranges are not currently supported in patterns, but they may be supported in
+the future.
+
+Syntactic ambiguity is not inherently bad. We see it every day in expressions like
+`a + b * c`. What is important is to disambiguate it reasonably by default and have a way to
+group operands in the alternative way when default disambiguation turns out to be incorrect.  
+In case of slice patterns the subslice interpretation seems overwhelmingly more likely, so we
+can take it as a default.  
+There was no visible demand for implementing half-bounded ranges in patterns so far, but if they
+are implemented in the future they will be able to be used in slice patterns as well, but they
+will require explicit grouping with recently implemented
+[parentheses in patterns](https://github.com/rust-lang/rust/pull/48500) (`[a, (..end)]`) or an
+explicitly written start boundary (`[a, 0 .. end]`).  
+We can also make *some* disambiguation effort and, for example, interpret `..LITERAL` as a
+range because `LITERAL` can never match a subslice. Time will show if such an effort is necessary
+or not.
+
+If/when half-bounded ranges are supported in patterns, for better future compatibility we'll need
+to reserve `..PAT` as "rest of the list" in tuples and tuple structs as well, and avoid interpreting
+it as a range pattern in those positions.
+
+Note that ambiguity with unbounded ranges as they are used in expressions (`..`) already exists in
+variant `Variant(..)` and tuple `(a, b, ..)` patterns, but it's very unlikely that the `..` syntax
+will ever be used in patterns in the range meaning because it duplicates functionality of the
+wildcard pattern `_`.
+
+#### `..PAT` vs `PAT..`
+
+Originally Rust used syntax `..PAT` for subslice patterns.  
+In 2014 the syntax was changed to `PAT..` by [RFC 202](https://github.com/rust-lang/rfcs/pull/202).
+That RFC received almost no discussion before it got merged and its motivation is no longer
+relevant because arrays now use syntax `[T; N]` instead of `[T, ..N]` used in old Rust.
+
+This RFC originally proposed to switch back to `..PAT`.
+Some reasons to switch:
+- Symmetry with expressions.  
+One of the general ideas behind patterns is that destructuring with
+patterns has the same syntax as construction with expressions, if possible.  
+In expressions we already have something with the meaning "rest of the list" - functional record
+update in struct expressions `S { field1, field2, ..remaining_fields }`.
+Right now we can use `S { field1, field1, .. }` in a pattern, but can't bind the remaining fields
+as a whole (by creating a new struct type on the fly, for example). It's not inconceivable that
+in Rust 2525 we have such ability and it's reasonable to expect it using syntax `..remaining_fields`
```

Comment:
What's the significance of 2525...?

Reply:
This was a [reference](https://www.youtube.com/watch?v=izQB2-Kmiic).